sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -0,0 +1,286 @@
+   * {@link SpannerIO#SPANNER_MUTATIONS_PER_COMMIT_LIMIT}. times). This means that the
+   * mutation operation should be idempotent.
+   */
+  @VisibleForTesting
+  static class SpannerWriterFn<T, V> extends DoFn<Mutation, Void> {
+    private static final Logger LOG = LoggerFactory.getLogger(SpannerWriterFn.class);
+    private transient Spanner spanner;
+    private final String projectId;
+    private final String instanceId;
+    private final String databaseId;
+    private final int batchSize;
+    private transient DatabaseClient dbClient;
+    // Current batch of mutations to be written.
+    private final List<Mutation> mutations = new ArrayList<>();
+
+    private static final int MAX_RETRIES = 5;
+    private static final FluentBackoff BUNDLE_WRITE_BACKOFF =
+        FluentBackoff.DEFAULT
+            .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+
+    @VisibleForTesting
+    SpannerWriterFn(String projectId, String instanceId, String databaseId, int batchSize) {
+      this.projectId = checkNotNull(projectId, "projectId");
+      this.instanceId = checkNotNull(instanceId, "instanceId");
+      this.databaseId = checkNotNull(databaseId, "databaseId");
+      this.batchSize = batchSize;
+    }
+
+    @Setup
+    public void setup() throws Exception {
+        SpannerOptions options = SpannerOptions.newBuilder().build();
+        spanner = options.getService();
+        dbClient = getDbClient(spanner, DatabaseId.of(projectId, instanceId, databaseId));
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      Mutation m = c.element();
+      mutations.add(m);
+      int columnCount = m.asMap().size();
+      if ((mutations.size() + 1) * columnCount >= batchSize) {
+        flushBatch();
+      }
+    }
+
+    @FinishBundle
+    public void finishBundle(Context c) throws Exception {
+      if (!mutations.isEmpty()) {
+        flushBatch();
+      }
+    }
+
+    @Teardown
+    public void teardown() throws Exception {
+      if (spanner == null) {
+          return;
+      }
+      spanner.closeAsync().get();
+    }
+
+    /**
+     * Writes a batch of mutations to Cloud Spanner.
+     *
+     * <p>If a commit fails, it will be retried up to {@link #MAX_RETRIES} times.
+     * If the retry limit is exceeded, the last exception from Cloud Spanner will be
+     * thrown.
+     *
+     * @throws SpannerException if the commit fails or IOException or InterruptedException if
+     * backing off between retries fails.
+     */
+    private void flushBatch() throws SpannerException, IOException, InterruptedException {
+      LOG.debug("Writing batch of {} mutations", mutations.size());
+      Sleeper sleeper = Sleeper.DEFAULT;
+      BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+
+      while (true) {
+        // Batch upsert rows.
+        try {
+          dbClient.writeAtLeastOnce(mutations);
```

Comment:
Spanner client library retries internally so I dont think you need to retry here for all SpannerExceptions. I think you can just catch AbortedException and retry only in that case.